### PR TITLE
Enhance Tensor indexSelect performance

### PR DESCRIPTION
**Pull request by JianpingChen066**
This is try to reduce the overhead on the index_select on CPU path at DLRM (https://github.com/facebookresearch/dlrm). To make src as contiguous can make it go into the parallelied path in Tensor indexSelect function

### CODE REVIEW

**Comment by JianpingChen066** [comment]
@VitalyFedyunin @fmassa @jgong5 please have a review on this . Thanks

**Comment by fmassa** [comment]
Can you share some representative sizes of the tensors that are used in this model?

Also, can you share some benchmark numbers before / after?

**Comment by cpuhrsch** [comment]
Is this for CPU or GPU?

**Comment by cpuhrsch** [comment]
@zou3519  - you might care about this as well since it's related to EmbeddingBag conceptually.

**Comment by JianpingChen066** [comment]
@fmassa 
For DLRM benchmark, run the benchmark test as below:
numactl --physcpubind=0-27 -m 0 python dlrm_s_pytorch.py --mini-batch-size=2048 --num-batches=1000 --data-generation=random --arch-mlp-bot=512-512-64 --arch-mlp-top=1024-1024-1024-1 --arch-sparse-feature-size=64 --arch-embedding-size=1000000-1000000-1000000-1000000-1000000-1000000-1000000-1000000 --num-indices-per-lookup=100 --arch-interaction-op=dot --numpy-rand-seed=727 --print-freq=100 --print-time --enable-profiling

And when run this benchmark, it's better to "export KMP_BLOCKTIME=1", then

without this patch , the CPU path benchmark result is as below by using the latest pytorch version (2ac9abf7591894525dfc7e76507f332df632ef19) :
Finished training it 100/1000 of epoch 0, 2969.25 ms/it, loss 0.220505, accuracy 0.000 %
Finished training it 200/1000 of epoch 0, 2997.01 ms/it, loss 0.085002, accuracy 0.000 %
Finished training it 300/1000 of epoch 0, 2971.44 ms/it, loss 0.084647, accuracy 0.000 %
Finished training it 400/1000 of epoch 0, 2976.13 ms/it, loss 0.084571, accuracy 0.000 %
Finished training it 500/1000 of epoch 0, 2972.50 ms/it, loss 0.084110, accuracy 0.000 %
Finished training it 600/1000 of epoch 0, 2977.38 ms/it, loss 0.084307, accuracy 0.000 %
Finished training it 700/1000 of epoch 0, 2976.58 ms/it, loss 0.084587, accuracy 0.000 %
Finished training it 800/1000 of epoch 0, 2980.39 ms/it, loss 0.084244, accuracy 0.000 %
Finished training it 900/1000 of epoch 0, 3008.62 ms/it, loss 0.084789, accuracy 0.000 %
Finished training it 1000/1000 of epoch 0, 2974.28 ms/it, loss 0.084187, accuracy 0.000 %

Apply this patch with same pytorch version, the result is as:
Finished training it 100/1000 of epoch 0, 1596.95 ms/it, loss 0.220505, accuracy 0.000 %
Finished training it 200/1000 of epoch 0, 1598.42 ms/it, loss 0.085002, accuracy 0.000 %
Finished training it 300/1000 of epoch 0, 1619.37 ms/it, loss 0.084647, accuracy 0.000 %
Finished training it 400/1000 of epoch 0, 1597.37 ms/it, loss 0.084571, accuracy 0.000 %
Finished training it 500/1000 of epoch 0, 1595.12 ms/it, loss 0.084110, accuracy 0.000 %
Finished training it 600/1000 of epoch 0, 1596.97 ms/it, loss 0.084307, accuracy 0.000 %
Finished training it 700/1000 of epoch 0, 1595.59 ms/it, loss 0.084587, accuracy 0.000 %
Finished training it 800/1000 of epoch 0, 1597.95 ms/it, loss 0.084244, accuracy 0.000 %
Finished training it 900/1000 of epoch 0, 1598.83 ms/it, loss 0.084789, accuracy 0.000 %
Finished training it 1000/1000 of epoch 0, 1601.12 ms/it, loss 0.084187, accuracy 0.000 %

Thanks

**Comment by Jianhui-Li** [comment]
@jspark1105 @dmudiger

**Comment by dmudiger** [comment]
cc: @bddpp (Junjie Bai), in reference to our earlier discussion

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp [hunk 1]
```diff
@@ -221,8 +221,12 @@ void THTensor_(indexSelect)(THTensor *tensor, THTensor *src, int dim, THLongTens
 
   index = THLongTensor_newContiguous(index);
   index_data = THLongTensor_data(index);
+  bool src_is_contiguous = src->is_contiguous();
+  if (!src_is_contiguous) {
+    src = THTensor_(newContiguous)(src);
+  }
 
-  if (dim == 0 && THTensor_(isContiguous)(src) && THTensor_(isContiguous)(tensor))
+  if (dim == 0 && THTensor_(isContiguous)(tensor))
```
Comment (VitalyFedyunin) [inline aten/src/TH/generic/THTensorEvenMoreMath.cpp]: You can get better performance by moving all ` src_is_contiguous` related code **into** this `if` branch.

Also feel free to simplify the code by:
```cpp
src = THTensor_(newContiguous)(src);
// do something with src
c10::raw::intrusive_ptr::decref(src);
```
as `THTensor_(newContiguous)(src)` doesn't copy already contiguous tensors.

Comment (JianpingChen066) [reply aten/src/TH/generic/THTensorEvenMoreMath.cpp]: Yes, code has been refined 
Thanks

**Review by VitalyFedyunin — DISMISSED** [review]

**Review by apaszke — COMMENTED** [review]
Are we sure this is always beneficial? While this might make it faster in some cases, for some (e.g. small/medium sized arrays that fit in cache) it's only likely to be slower...

**Comment by vincentqb** [comment]
> Are we sure this is always beneficial? While this might make it faster in some cases, for some (e.g. small/medium sized arrays that fit in cache) it's only likely to be slower...

Due to this comment, I'm holding on landing this PR.

**Comment by JianpingChen066** [comment]
@apaszke 
Hi, Adam
are there any existing test cases for small/medium sized arrays performance test thus I can have a check on this ? Or is there some heuristic limitation similar as TH_OMP_OVERHEAD_THRESHOLD thus I can used to set a threshold on this ?  
Thanks

**Comment by JianpingChen066** [comment]
Hi, Adam 
What do you think about I also use TH_OMP_OVERHEAD_THRESHOLD as a threshold to do newContiguous for src liek below:

  auto src_size0 = THTensor_sizeLegacyNoScalars(src, 0);
  ptrdiff_t rowsize = src_size0 == 0 ? 1 : THTensor_(nElement)(src) / src_size0;
  auto omp_threshold = TH_OMP_OVERHEAD_THRESHOLD;
  if (src->dim() > 1) {
    omp_threshold = TH_OMP_OVERHEAD_THRESHOLD / rowsize;
  }
  if(!THTensor_(isContiguous)(src) && (numel > omp_threshold)) {
    src = THTensor_(newContiguous)(src);
  }
  ...
  if(!THTensor_(isContiguous)(src) && (numel > omp_threshold)) {
    c10::raw::intrusive_ptr::decref(src);
  }

so this can ensure the parallel code must be executed and can get better performance than sequential code ?

Regards,
Jianping

**Comment by JianpingChen066** [comment]
@VitalyFedyunin @fmassa @jgong5 please have a further review on this . Thanks

**Comment by apaszke** [comment]
@JianpingChen066 It's hard in general to pick the right threshold for every occasion, because it depends on the interactions between multiple non-trivial CPU features. Can you please try to run this operation on tensors of varying sizes (with indices both coalesced/sorted and completely scattered over the range) and report the timings before and after the patch?

**Comment by JianpingChen066** [comment]
@apaszke 
Hi, Adam
Are there any existing test cases for indexSelect functions ? 
We only want src as contiguous as possible thus parallel version can work better than sequential version, it should have no relation with whether indices is coalesced or not, sorted or not and scattered or not ?

**Comment by ezyang** [comment]
Sorry I accidentally clicked land on this

**Comment by VitalyFedyunin** [comment]
@JianpingChen066 it is important NOT to use any OMP, TDD or similar references inside the code base as we try to make them interchangeable ( so please do not use `TH_OMP_OVERHEAD_THRESHOLD` ).

Also as @apaszke wrote, it would be nice to see benchmarks for small and large tensors before and after patch. You can find some index_select tests inside of the `def test_index_select(self)` as starting point. But you need to write your own benchmarking script.

**Comment by ezyang** [comment]
@VitalyFedyunin @ilia-cher Can we add some sort of lint for this rule? It's very easy to forget...

**Comment by VitalyFedyunin** [comment]
PR reverted from the master. You can rebase and keep working on it.

**Comment by JianpingChen066** [comment]
@VitalyFedyunin Yes, while TH_OMP_OVERHEAD_THRESHOLD is in the previous code and defined at THTensorApply.hpp. It is used by the indexSelect in the at::parallel_for as parallelization threshold. May I change it to TH_PARALLEL_FOR_OVERHEAD_THRESHOLD ?  Thanks

**Comment by JianpingChen066** [comment]
@VitalyFedyunin @apaszke 
I have write one test case with different index sizes and combined with src is contiguous or not as below:
 import time
import random
import torch

from common_utils import TestCase, run_tests
from hypothesis import given, settings
from hypothesis import strategies as st

class indexSelectTest(TestCase):
  @settings(deadline=None)
  @given(batch_size=st.sampled_from([10000000, 1000, 10000, 1000000]))
  def test_index_select(self, batch_size):
    print('\ntest_index_select:')
    random.seed(100)
    index = random.sample(range(batch_size), batch_size//10)
    src = torch.randn(batch_size, 64, device='cpu')
    idx = torch.tensor(index, dtype=torch.long, device='cpu')
    start = time.time()
    for i in range(100):
      dst = torch.index_select(src, 0, idx)
    end = time.time()
    print('index size:{}, time: {:.6f} ms/per_index_select'.format(len(index), (end-start)*10))

  @settings(deadline=None)
  @given(batch_size=st.sampled_from([10000000, 1000, 10000, 1000000]))
  def test_index_select_non_contiguous(self, batch_size):
    print('\ntest_index_select_non_contiguous:')
    random.seed(100)
    index = random.sample(range(batch_size), batch_size//10)
    src = torch.randn(batch_size, 64, 2, device='cpu')
    non_contiguous_src = src.select(2, 0)
    idx = torch.tensor(index, dtype=torch.long, device='cpu')
    start = time.time()
    for i in range(100):
      dst = torch.index_select(non_contiguous_src, 0, idx)
    end = time.time()
    print('index size:{}, time: {:.6f} ms/per_index_select'.format(len(index), (end-start)*10))

if __name__=='__main__':
  run_tests()


The result for the original indexSelect function is like below:
**test_index_select:**
index size:1000000, time: 65.922041 ms/per_index_select
index size:100000, time: 6.105187 ms/per_index_select
index size:1000, time: 0.055368 ms/per_index_select
index size:100, time: 0.005496 ms/per_index_select
**test_index_select_non_contiguous:**
index size:1000000, time: 2191.913469 ms/per_index_select
index size:100000, time: 213.450601 ms/per_index_select
index size:1000, time: 1.596482 ms/per_index_select
index size:100, time: 0.161395 ms/per_index_select

The result for this patch is like:
**test_index_select:**
index size:1000000, time: 66.463988 ms/per_index_select
index size:100000, time: 6.164207 ms/per_index_select
index size:1000, time: 0.056939 ms/per_index_select
index size:100, time: 0.005827 ms/per_index_select
**test_index_select_non_contiguous:**
index size:1000000, time: 750.100091 ms/per_index_select
index size:100000, time: 72.395074 ms/per_index_select
index size:1000, time: 1.787453 ms/per_index_select
index size:100, time: 0.172462 ms/per_index_select

The small size may have some impact while almost can be ignored and the bigger size with non-contiguous src (large than 100000) can get almost 3x improvement

May I also need to make test case to be checked in ?
Thanks

**Comment by fmassa** [comment]
Can you also add a test case to your benchmark where the number of indexed elements is different from the dimension being indexed? Similar tohttps://github.com/pytorch/pytorch/pull/13420
For example, with 1024×1024 -> 128×1024 ?

**Comment by JianpingChen066** [comment]
@fmassa 
Hi Francisco 
I modified codes a little to make it has less impact when dim is not 0
and add some test cases into the test_index_select.py
The result is like below:
                                     |    **Old**          |    **New**          |
                                       ms/per_index_select
**test_index_select:** 
index   size:1000000,   | 40.03756      | 40.477538   |
index   size:100000,     | 5.590582      | 5.971432     |
index   size:1000,         | 0.078225      | 0.077724     |
index   size:100,           | 0.007057      | 0.006747     |
**test_index_select_non_contiguous:**  
index   size:1000000,   | 2231.867933 | 427.461815 |
index   size:100000,     | 240.299988   | 213.389063 |
index   size:1000,         | 1.578188       | 1.731372     |
index   size:100,           | 0.167203       | 0.173616     |
**test_index_select_non_contiguous_dim0 :**  
index   size:128,           | 0.341105       | 0.34554       |
index   size:131072,     | 891.35623     | 771.395373 |
index   size:16384,       | 111.90021     | 111.182475 |
index   size:512,           | 1.488233       | 1.529646     |
**test_index_select_non_contiguous_dim1:**   
index   size:128,           | 0.943565       | 0.925708     |
index   size:131072,     | 6306.76558   | 6298.30513 |
index   size:16384,       | 447.111034   | 445.347142 |
index   size:512,           | 7.793069       | 7.451367     |

File: aten/src/TH/generic/THTensorApply.hpp [hunk 2]
```diff
@@ -8,6 +8,7 @@
 #define ORDIN_TH_OMP_OVERHEAD_THRESHOLD 20000
 #define UNCERTAIN_TH_OMP_OVERHEAD_THRESHOLD 50000
 #define TH_OMP_OVERHEAD_THRESHOLD 100000
+#define TH_PARALLEL_OVERHEAD_THRESHOLD 80000
```
Comment (VitalyFedyunin) [inline aten/src/TH/generic/THTensorApply.hpp]: Please move it into the THTensorEvenMoreMath.cpp as constexpr, we are trying to get rid of THTensorApply.
Also something like `parallel_overhead_threshold_index_select` would be better as name ( + comment how we selected this `80000` value)

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Overall good, but you can get MUCH better performance if you move this operator to aten (see examples in #24507

**Comment by ifedan** [comment]
> Overall good, but you can get MUCH better performance if you move this operator to aten (see examples in #24507

I agree with @VitalyFedyunin , I think it's a good candidate to be moved from TH to ATen and to start using TensorIterator

**Comment by JianpingChen066** [comment]
@ifedan @VitalyFedyunin 
We are working on this to move indexSelect from TH to ATen, Thanks

**Comment by JianpingChen066** [comment]
@ifedan @VitalyFedyunin 
Mingfei is improving the index_select performance and move it to  aten/src/ATen/native/Indexing.cpp, please refer his [PR#30598](https://github.com/pytorch/pytorch/pull/30598). I will close this PR now.

Thanks
